src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -291,69 +326,127 @@ private async Task<string> DownloadAsync(Uri zipUri)
+                }
+                else
                 {
-                    _logger.LogInformation($"Extracting files to '{scriptPath}'");
-                    ZipFile.ExtractToDirectory(filePath, scriptPath, overwriteFiles: true);
-                    _logger.LogInformation($"Zip extraction complete");
+                    UnzipPackage(filePath, scriptPath);
                 }
             }
+        }
+
+        private CodePackageType GetPackageType(string filePath)
+        {
+            // Check file magic-number using `file` command.
+            (var output, _, _) = RunBashCommand($"file -b {filePath}", MetricEventNames.LinuxContainerSpecializationFileCommand);
+            if (output.StartsWith("Squashfs", StringComparison.OrdinalIgnoreCase))
+            {
+                return CodePackageType.Squashfs;
+            }
+            else if (output.StartsWith("Zip", StringComparison.OrdinalIgnoreCase))
+            {
+                return CodePackageType.Zip;
+            }
+            else if (FileIsAny(".squashfs", ".sfs", ".sqsh", ".img", ".fs"))
+            {
+                return CodePackageType.Squashfs;
+            }
+            else if (FileIsAny(".zip"))
+            {
+                return CodePackageType.Zip;
+            }
+            else
+            {
+                throw new InvalidOperationException($"Can't find CodePackageType to match {filePath}");
+            }
 
             bool FileIsAny(params string[] options)
                 => options.Any(o => filePath.EndsWith(o, StringComparison.OrdinalIgnoreCase));
         }
 
-        private void MountFsImage(string filePath, string scriptPath)
+        private void UnzipPackage(string filePath, string scriptPath)
+        {
+            using (_metricsLogger.LatencyEvent(MetricEventNames.LinuxContainerSpecializationZipExtract))
+            {
+                _logger.LogInformation($"Extracting files to '{scriptPath}'");
+                ZipFile.ExtractToDirectory(filePath, scriptPath, overwriteFiles: true);
+                _logger.LogInformation($"Zip extraction complete");
+            }
+        }
+
+        private void UnsquashImage(string filePath, string scriptPath)
+            => RunBashCommand($"(mkdir -p '{scriptPath}' || true) && unsquashfs -f -d '{scriptPath}' '{filePath}'", MetricEventNames.LinuxContainerSpecializationUnsquash);
+
+        private void MountSquashfsImage(string filePath, string scriptPath)
             => RunFuseMount($"squashfuse_ll '{filePath}' '{scriptPath}'", scriptPath);
 
         private void MountZipFile(string filePath, string scriptPath)
             => RunFuseMount($"fuse-zip -r '{filePath}' '{scriptPath}'", scriptPath);
 
         private void RunFuseMount(string mountCommand, string targetPath)
+            => RunBashCommand($"(mknod /dev/fuse c 10 229 || true) && (mkdir -p '{targetPath}' || true) && ({mountCommand})", MetricEventNames.LinuxContainerSpecializationFuseMount);
```

Comment:
May want to change the command into `(mknod /dev/fuse c 10 229 || true) && (mkdir -p '{targetPath}' || true) && **(rm -rf '{targetPath}/*' || true)** && ({mountCommand})`.

This is to deal with the situation https://github.com/Azure/azure-functions-docker/pull/117, where the `start.sh` in python worker will create `/home/site/wwwroot/.python_packages`.

Reply:
we can just tell squashfuse_ll and fuse-zip to ignore non-empty mountpoints